File: internal/plugins/jest/rules/prefer_spy_on/prefer_spy_on_test.go
package prefer_spy_on_test

import (
	"testing"

	"github.com/web-infra-dev/rslint/internal/plugins/jest/fixtures"
	"github.com/web-infra-dev/rslint/internal/plugins/jest/rules/prefer_spy_on"
	"github.com/web-infra-dev/rslint/internal/rule_tester"
)

func TestPreferSpyOnRule(t *testing.T) {
	rule_tester.RunRuleTester(
		fixtures.GetRootDir(),
		"tsconfig.json",
		t,
		&prefer_spy_on.PreferSpyOnRule,
		[]rule_tester.ValidTestCase{
			{Code: `Date.now = () => 10`},
			{Code: `window.fetch = jest.fn`},
			{Code: `Date.now = fn()`},
			{Code: `obj.mock = jest.something()`},
			{Code: `const mock = jest.fn()`},
			{Code: `mock = jest.fn()`},
			{Code: `const mockObj = { mock: jest.fn() }`},
			{Code: `mockObj = { mock: jest.fn() }`},
			{Code: "window[`${name}`] = jest[`fn${expression}`]()"},
		},
		[]rule_tester.InvalidTestCase{
			{
				Code:   `obj.a = jest.fn(); const test = 10;`,
				Output: []string{`jest.spyOn(obj, 'a').mockImplementation(); const test = 10;`},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "useJestSpyOn", Line: 1, Column: 1, EndLine: 1, EndColumn: 18},
				},
			},
			{
				Code:   `Date['now'] = jest['fn']()`,
				Output: []string{`jest.spyOn(Date, 'now').mockImplementation()`},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "useJestSpyOn", Line: 1, Column: 1, EndLine: 1, EndColumn: 27},
				},
			},
			{
				Code:   "window[`${name}`] = jest[`fn`]()",
				Output: []string{"jest.spyOn(window, `${name}`).mockImplementation()"},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "useJestSpyOn", Line: 1, Column: 1, EndLine: 1, EndColumn: 33},
				},
			},
			{
				Code:   `obj['prop' + 1] = jest['fn']()`,
				Output: []string{`jest.spyOn(obj, 'prop' + 1).mockImplementation()`},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "useJestSpyOn", Line: 1, Column: 1, EndLine: 1, EndColumn: 31},
				},
			},
			{
				Code:   `obj.one.two = jest.fn(); const test = 10;`,
				Output: []string{`jest.spyOn(obj.one, 'two').mockImplementation(); const test = 10;`},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "useJestSpyOn", Line: 1, Column: 1, EndLine: 1, EndColumn: 24},
				},
			},
			{
				Code:   `obj.a = jest.fn(() => 10,)`,
				Output: []string{`jest.spyOn(obj, 'a').mockImplementation(() => 10)`},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "useJestSpyOn", Line: 1, Column: 1, EndLine: 1, EndColumn: 27},
				},
			},
			{
				Code:   `obj.a.b = jest.fn(() => ({})).mockReturnValue('default').mockReturnValueOnce('first call'); test();`,
				Output: []string{`jest.spyOn(obj.a, 'b').mockImplementation(() => ({})).mockReturnValue('default').mockReturnValueOnce('first call'); test();`},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "useJestSpyOn", Line: 1, Column: 1, EndLine: 1, EndColumn: 91},
				},
			},
			{
				Code:   `window.fetch = jest.fn(() => ({})).one.two().three().four`,
				Output: []string{`jest.spyOn(window, 'fetch').mockImplementation(() => ({})).one.two().three().four`},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "useJestSpyOn", Line: 1, Column: 1, EndLine: 1, EndColumn: 58},
				},
			},
			{
				Code:   `foo[bar] = jest.fn().mockReturnValue(undefined)`,
				Output: []string{`jest.spyOn(foo, bar).mockImplementation().mockReturnValue(undefined)`},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "useJestSpyOn", Line: 1, Column: 1, EndLine: 1, EndColumn: 48},
				},
			},
			{
				Code: `
        foo.bar = jest.fn().mockImplementation(baz => baz)
        foo.bar = jest.fn(a => b).mockImplementation(baz => baz)
      `,
				Output: []string{`
        jest.spyOn(foo, 'bar').mockImplementation(baz => baz)
        jest.spyOn(foo, 'bar').mockImplementation(baz => baz)
      `},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "useJestSpyOn", Line: 2, Column: 9, EndLine: 2, EndColumn: 59},
					{MessageId: "useJestSpyOn", Line: 3, Column: 9, EndLine: 3, EndColumn: 65},
				},
			},
			{
				Code:   `foo.bar = (jest.fn()).mockImplementation(baz => baz)`,
				Output: []string{`jest.spyOn(foo, 'bar').mockImplementation(baz => baz)`},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "useJestSpyOn", Line: 1, Column: 1, EndLine: 1, EndColumn: 53},
				},
			},
			{
				Code:   `foo.bar = (jest.fn(a => b)).mockImplementation(baz => baz)`,
				Output: []string{`jest.spyOn(foo, 'bar').mockImplementation(baz => baz)`},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "useJestSpyOn", Line: 1, Column: 1, EndLine: 1, EndColumn: 59},
				},
			},
			{
				Code:   `obj.a = (jest.fn())`,
				Output: []string{`jest.spyOn(obj, 'a').mockImplementation()`},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "useJestSpyOn", Line: 1, Column: 1, EndLine: 1, EndColumn: 20},
				},
			},
			{
				Code:   `obj.a = ((jest.fn()))`,
				Output: []string{`jest.spyOn(obj, 'a').mockImplementation()`},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "useJestSpyOn", Line: 1, Column: 1, EndLine: 1, EndColumn: 22},
				},
			},
			{
				Code:   `obj.a = (jest.fn(() => 10))`,
				Output: []string{`jest.spyOn(obj, 'a').mockImplementation(() => 10)`},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "useJestSpyOn", Line: 1, Column: 1, EndLine: 1, EndColumn: 28},
				},
			},
			{
				Code:   `obj.a = (jest.fn()).one.two()`,
				Output: []string{`jest.spyOn(obj, 'a').mockImplementation().one.two()`},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "useJestSpyOn", Line: 1, Column: 1, EndLine: 1, EndColumn: 30},
				},
			},
		},
	)
}

File: packages/rslint-test-tools/tests/rstest/rules/prefer-spy-on.test.ts
import { RuleTester } from '../rule-tester';

const ruleTester = new RuleTester();

ruleTester.run('prefer-spy-on', {} as never, {
  valid: [
    { code: 'Date.now = () => 10' },
    { code: 'window.fetch = rs.fn' },
    { code: 'Date.now = fn()' },
    { code: 'obj.mock = rs.something()' },
    { code: 'const mock = rs.fn()' },
    { code: 'mock = rs.fn()' },
    { code: 'const mockObj = { mock: rs.fn() }' },
    { code: 'mockObj = { mock: rs.fn() }' },
    { code: 'window[`${name}`] = rs[`fn${expression}`]()' },
  ],
  invalid: [
    {
      code: 'obj.a = rs.fn(); const test = 10;',
      output: `rs.spyOn(obj, 'a').mockImplementation(() => undefined); const test = 10;`,
      errors: [
        {
          messageId: 'useRsSpyOn',
          line: 1,
          column: 1,
          endLine: 1,
          endColumn: 16,
        },
      ],
    },
    {
      code: `Date['now'] = rs['fn']()`,
      output: `rs.spyOn(Date, 'now').mockImplementation(() => undefined)`,
      errors: [
        {
          messageId: 'useRsSpyOn',
          line: 1,
          column: 1,
          endLine: 1,
          endColumn: 25,
        },
      ],
    },
    {
      code: 'window[`${name}`] = rs[`fn`]()',
      output: 'rs.spyOn(window, `${name}`).mockImplementation(() => undefined)',
      errors: [
        {
          messageId: 'useRsSpyOn',
          line: 1,
          column: 1,
          endLine: 1,
          endColumn: 31,
        },
      ],
    },
    {
      code: `obj['prop' + 1] = rs['fn']()`,
      output: `rs.spyOn(obj, 'prop' + 1).mockImplementation(() => undefined)`,
      errors: [
        {
          messageId: 'useRsSpyOn',
          line: 1,
          column: 1,
          endLine: 1,
          endColumn: 29,
        },
      ],
    },
    {
      code: 'obj.one.two = rs.fn(); const test = 10;',
      output: `rs.spyOn(obj.one, 'two').mockImplementation(() => undefined); const test = 10;`,
      errors: [
        {
          messageId: 'useRsSpyOn',
          line: 1,
          column: 1,
          endLine: 1,
          endColumn: 22,
        },
      ],
    },
    {
      code: 'obj.a = rs.fn(() => 10,)',
      output: `rs.spyOn(obj, 'a').mockImplementation(() => 10)`,
      errors: [
        {
          messageId: 'useRsSpyOn',
          line: 1,
          column: 1,
          endLine: 1,
          endColumn: 25,
        },
      ],
    },
    {
      code: `obj.a.b = rs.fn(() => ({})).mockReturnValue('default').mockReturnValueOnce('first call'); test();`,
      output: `rs.spyOn(obj.a, 'b').mockImplementation(() => ({})).mockReturnValue('default').mockReturnValueOnce('first call'); test();`,
      errors: [
        {
          messageId: 'useRsSpyOn',
          line: 1,
          column: 1,
          endLine: 1,
          endColumn: 89,
        },
      ],
    },
    {
      code: 'window.fetch = rs.fn(() => ({})).one.two().three().four',
      output: `rs.spyOn(window, 'fetch').mockImplementation(() => ({})).one.two().three().four`,
      errors: [
        {
          messageId: 'useRsSpyOn',
          line: 1,
          column: 1,
          endLine: 1,
          endColumn: 56,
        },
      ],
    },
    {
      code: 'foo[bar] = rs.fn().mockReturnValue(undefined)',
      output:
        'rs.spyOn(foo, bar).mockImplementation(() => undefined).mockReturnValue(undefined)',
      errors: [
        {
          messageId: 'useRsSpyOn',
          line: 1,
          column: 1,
          endLine: 1,
          endColumn: 46,
        },
      ],
    },
    {
      code: `
        foo.bar = rs.fn().mockImplementation(baz => baz)
        foo.bar = rs.fn(a => b).mockImplementation(baz => baz)
      `,
      output: `
        rs.spyOn(foo, 'bar').mockImplementation(baz => baz)
        rs.spyOn(foo, 'bar').mockImplementation(baz => baz)
      `,
      errors: [
        {
          messageId: 'useRsSpyOn',
          line: 2,
          column: 9,
          endLine: 2,
          endColumn: 57,
        },
        {
          messageId: 'useRsSpyOn',
          line: 3,
          column: 9,
          endLine: 3,
          endColumn: 63,
        },
      ],
    },
  ],
});
